dali/operators/image/resize/resize_attr.cc

```diff
@@ -0,0 +1,355 @@
+  params_.resize(N);
+
+  if (has_size_) {
+    GetShapeLikeArgument<float>(size_arg_, spec, "size", ws, spatial_ndim_, N);
+  }
+
+  max_size_.resize(spatial_ndim_, std::numeric_limits<int>::max());
+  if (has_max_size_) {
+    GetSingleOrRepeatedArg(spec, max_size_, "max_size", spatial_ndim_);
+  }
+
+  SmallVector<float, 3> requested_size, in_size;
+  requested_size.resize(spatial_ndim_);
+  in_size.resize(spatial_ndim_);
+
+  if (HasSeparateSizeArgs()) {
+    res_x_.resize(N);
+    res_y_.resize(N);
+    res_z_.resize(N);
+
+    float *size_vecs[3] = { nullptr, nullptr, nullptr };
+    {
+      // outer-first shape order :(
+      int d = 0;
+      if (spatial_ndim_ >= 3)
+        size_vecs[d++] = res_z_.data();
+      if (spatial_ndim_ >= 2)
+        size_vecs[d++] = res_y_.data();
+      size_vecs[d++] = res_x_.data();
+    }
+
+    if (has_resize_x_) {
+      GetPerSampleArgument(res_x_, "resize_x", spec, ws, N);
+    }
+
+    if (has_resize_y_) {
+      GetPerSampleArgument(res_y_, "resize_y", spec, ws, N);
+    }
+
+    if (has_resize_z_) {
+      GetPerSampleArgument(res_z_, "resize_z", spec, ws, N);
+    }
+
+    for (int i = 0; i < N; i++) {
+      for (int d = 0; d < spatial_ndim_; d++) {
+        in_size[d] = input_shape.tensor_shape_span(i)[d + first_spatial_dim_];
+        requested_size[d] = size_vecs[d][i];
+      }
+
+      CalculateSampleParams(params_[i], requested_size, in_size, subpixel_scale_);
+    }
+  } else if (has_resize_shorter_ || has_resize_longer_) {
+    const char *arg_name = has_resize_shorter_ ? "resize_shorter" : "resize_longer";
+    GetPerSampleArgument(res_x_, arg_name, spec, ws, N);
+
+    for (int i = 0; i < N; i++) {
+      for (int d = 0; d < spatial_ndim_; d++) {
+        in_size[d] = input_shape.tensor_shape_span(i)[d + first_spatial_dim_];
+        requested_size[d] = res_x_[i];
+      }
+
+      CalculateSampleParams(params_[i], requested_size, in_size, subpixel_scale_);
+    }
+  } else if (has_size_) {
+    for (int i = 0; i < N; i++) {
+      auto in_sample_shape = input_shape.tensor_shape_span(i);
+      for (int d = 0; d < spatial_ndim_; d++) {
+        in_size[d] = in_sample_shape[d + first_spatial_dim_];
+        requested_size[d] = size_arg_[i * spatial_ndim_ + d];
+      }
+
+      CalculateSampleParams(params_[i], requested_size, in_size, subpixel_scale_);
+    }
+  }
+}
+
+void ResizeAttr::AdjustOutputSize(float *out_size, const float *in_size) {
+  SmallVector<double, 3> scale;
+  SmallVector<bool, 3> mask;
+  scale.resize(spatial_ndim_, 1);
+  mask.resize(spatial_ndim_);
+
+  int sizes_provided = 0;
+  for (int d = 0; d < spatial_ndim_; d++) {
+    mask[d] = (out_size[d] != 0);
+    scale[d] = in_size[d] ? out_size[d] / in_size[d] : 1;
+    sizes_provided += mask[d];
+  }
+
+  if (sizes_provided == 0) {  // no clue how to resize - keep original size then
+    for (int d = 0; d < spatial_ndim_; d++)
+      out_size[d] = in_size[d];
+  } else {
+    if (mode_ == ResizeMode::Default) {
```

Comment:
It would be good to move every case: `ResizeMode::Default`, `mode_ == ResizeMode::Stretch` and the rest to a separate functions. Now it is a bit too long to read in one go.

Reply:
On the other hand, these functions would have ~7 parameters if I were to forward all necessary local variables.